### PR TITLE
Refactoring Of The Checkout Controller

### PR DESCRIPTION
**Pull request by dvassev**
This refactoring decouples the update method, removing state leakage, and simplification of the method overall.

### CODE REVIEW

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 1]
```diff
@@ -25,33 +25,61 @@ class CheckoutController < Spree::StoreController
...
-        end
-        if !success
-          flash[:error] = @order.errors.full_messages.join("\n")
-          redirect_to(checkout_state_path(@order.state)) && return
-        end
+      if order_update
+
+        assign_temp_address
+
+        on_failure and return unless state_transition
```
Comment (mtomov) [inline frontend/app/controllers/spree/checkout_controller.rb]: You might want to consider renaming `order_update`, `state_transition` and `on_failure` to their verb first equivalents to follow the schematics on the rest of the methods and to indicate that an action is happening. Say to `update_order`, `transition_forward!` and `redirect_on_failure`

**Comment by dvassev** [comment]
@mtomov Thank you very much for this. I didn't notice that we you guys were using verb first and it makes a ton of sense.

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 2]
```diff
@@ -25,33 +25,62 @@ class CheckoutController < Spree::StoreController
...
+    def assign_temp_address
+      @order.temporary_address = !params[:save_user_address]
+    end
+
+    def redirect_on_failure
+      flash[:error] = @order.errors.full_messages.join("\n")
+      redirect_to(checkout_state_path(@order.state))
+    end
+
+    def transition_forward!
```
Comment (jhawthorn) [inline frontend/app/controllers/spree/checkout_controller.rb]: I don't think this should be a "bang" methd. There's no non-bang version of this. We're calling `next` and `complete` on order, not `next!` and `complete!`.

I'd prefer it was just called `transition_forward` or similar.

File: frontend/app/controllers/spree/checkout_controller.rb [hunk 3]
```diff
@@ -25,33 +25,62 @@ class CheckoutController < Spree::StoreController
...
-        end
-        if !success
-          flash[:error] = @order.errors.full_messages.join("\n")
-          redirect_to(checkout_state_path(@order.state)) && return
-        end
+      if update_order
+
+        assign_temp_address
+
+        redirect_on_failure and return unless transition_forward!
```
Comment (jhawthorn) [inline frontend/app/controllers/spree/checkout_controller.rb]: I find this line tough to parse for a couple reasons.

I prefer using the conditional suffix only on functions that don't mutate state. It's kind of sneaky that `transition_forward!`, the main purpose of this whole action, is at the end of the line. It also makes the order these run in mildly confusing: `this_second and this_third unless this_first`.


I'd prefer that the `transition_forward!` was more obvious, and that the order would read top-to-bottom

``` ruby
unless transition_forward!
  redirect_on_failure
  return
end
```

Comment (dvassev) [inline frontend/app/controllers/spree/checkout_controller.rb]: It is definitely more readable and more accessible even for people coming from other languages.

**Comment by dvassev** [comment]
@jhawthorn Thank you for the suggestions.

**Pull request merged**